include/dali/core/mm/async_pool.h

```diff
@@ -0,0 +1,428 @@
+        size_t split_size = block_size;
+        if (split) {
+          // Adjust the pending free `f` so that it contains only what remains after
+          // the block was split.
+          size_t remainder_size = block_end - remainder;
+          f->addr = remainder;
+          f->bytes = remainder_size;
+          f->alignment = remainder_alignment;
+          from.by_size.insert({ remainder_size, f });
+          // The block taken out from the free list is now reduced
+          split_size = remainder - base;
+        } else {
+          remove_pending_free(from, f, false);
+        }
+        if (split_size != bytes) {
+          padded_[aligned] = { split_size,
+                               static_cast<int>(front_padding),
+                               static_cast<int>(alignment) };
+        }
+        return aligned;
+      }
+    }
+    return nullptr;
+  }
+
+  /**
+   * @brief Searches per-stream free blocks to find the most recently freed one.
+   */
+  pending_free *find_first_ready(PerStreamFreeBlocks &free) {
+    SmallVector<pending_free *, 128> pending;
+    int step = 1;
+    pending_free *f = free.free_list.head;
+    while (f) {
+      if (f->ready())
+        break;
+      pending.clear();
+      for (int i = 0; i < step; i++) {
+        f = f->next;
+        if (!f)
+          break;
+        pending.push_back(f);
+      }
+    }
+    if (pending.empty()) {
+      if (f) {
+        assert(f->ready());
+        assert(!f->prev || !f->ready());
+      }
+      return f;
+    }
+    auto it = std::partition_point(pending.begin(), pending.end(), [&](pending_free *f) {
+      return !f->ready();
+    });
+    if (it == pending.end()) {
+      assert(!free.free_list.tail || !free.free_list.tail->ready());
+      return nullptr;
+    }
+    f = *it;
+    assert(f->ready());
+    return f;
+  }
+
+  void free_pending(PerStreamFreeBlocks &free) {
+    auto *f = find_first_ready(free);
+    while (f) {
+      global_pool_.deallocate(f->addr, f->bytes, f->alignment);
+      f = remove_pending_free(free, f);
+    }
+  }
+
+  void deallocate_async_impl(PerStreamFreeBlocks &free, char *ptr, size_t bytes, size_t alignment,
+                             cudaStream_t stream) {
+    auto *pending = add_pending_free(free.free_list, ptr, bytes, alignment, stream);
+    try {
+      free.by_size.insert({bytes, pending});
+    } catch (...) {
+      remove_pending_free(free.free_list, pending);
+      throw;
+    }
+  }
+
+  /**
+   * @brief Restores original block parameters (address, size and alignment) of a block
+   *        that was used for allocating a smaller block, possibly with stricter alignment.
+   */
+  void restore_original_params(char *&p, size_t &bytes, size_t &alignment) {
+    auto it = padded_.find(p);
+    if (it != padded_.end()) {
+      if (it->second.front_padding + bytes > it->second.bytes) {
+        throw std::invalid_argument("The deallocated memory points to a block that's smaller than "
```

Comment:
This error message clearly states that this function can be used only for deallocation, while the name of it is very generic.
I would rename the function somehow.

Reply:
There are more problems with this function - the biggest being that it also removes the pointer from `padded_`. I have no idea what better name could there be without getting ridiculously long. The real meaning is:
`restore_and_erase_original_allocation_params_before_deallocation` which is ludicrous.

Reply:
Also, this should probably be just an assert.